targets/TARGET_ARM_SSG/TARGET_MUSCA_A1/us_ticker.c

```diff
@@ -0,0 +1,141 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * Supports the High-resolution Ticker for mbed by implementing
+ * \ref us_ticker_api.h, using a CMSDK Timer \ref timer_cmsdk_dev_t.
+ */
+
+#include "device.h"
+#include "timer_cmsdk_drv.h"
+#include "us_ticker_api.h"
+
+#if TARGET_MUSCA_A1_NS
+
+static uint64_t total_ticks = 0;
+static uint32_t previous_ticks = 0;
+
+static void restart_timer(uint32_t new_reload)
+{
+    timer_cmsdk_disable(&USEC_TIMER_DEV);
+    timer_cmsdk_set_reload_value(&USEC_TIMER_DEV,
+                                 new_reload);
+    timer_cmsdk_reset(&USEC_TIMER_DEV);
+    timer_cmsdk_clear_interrupt(&USEC_TIMER_DEV);
+    timer_cmsdk_enable_interrupt(&USEC_TIMER_DEV);
+    timer_cmsdk_enable(&USEC_TIMER_DEV);
+}
+
+void us_ticker_init(void)
+{
+    timer_cmsdk_init(&USEC_TIMER_DEV);
+    previous_ticks = TIMER_CMSDK_MAX_RELOAD;
+    NVIC_EnableIRQ(USEC_INTERVAL_IRQ);
+    restart_timer(previous_ticks);
+}
+
+void us_ticker_free(void)
+{
+    timer_cmsdk_disable(&USEC_TIMER_DEV);
+}
+
+uint32_t us_ticker_read(void)
+{
+    if (timer_cmsdk_is_interrupt_active(&USEC_TIMER_DEV)) {
+        total_ticks += previous_ticks;
+        previous_ticks = TIMER_CMSDK_MAX_RELOAD;
+        restart_timer(previous_ticks);
+    }
+    uint32_t tick = timer_cmsdk_get_current_value(&USEC_TIMER_DEV);
+
+    if (tick < previous_ticks) {
+        uint32_t delta = previous_ticks - tick;
+        total_ticks += delta;
+        previous_ticks = tick;
+    }
+
+    return (total_ticks >> USEC_REPORTED_SHIFT);
+}
+
+void us_ticker_set_interrupt(timestamp_t timestamp)
+{
+    uint32_t reload = (timestamp - us_ticker_read()) << USEC_REPORTED_SHIFT;
+    previous_ticks = reload;
+    restart_timer(previous_ticks);
+}
+
+void us_ticker_disable_interrupt(void)
+{
+    timer_cmsdk_disable_interrupt(&USEC_TIMER_DEV);
+}
+
+void us_ticker_clear_interrupt(void)
+{
+    timer_cmsdk_clear_interrupt(&USEC_TIMER_DEV);
+}
+
+void us_ticker_fire_interrupt(void)
+{
+    NVIC_SetPendingIRQ(USEC_INTERVAL_IRQ);
+}
+
+const ticker_info_t* us_ticker_get_info()
+{
+    static const ticker_info_t info = {
+        USEC_REPORTED_FREQ_HZ,
+        USEC_REPORTED_BITS
+    };
+    return &info;
+}
+
+#ifndef usec_interval_irq_handler
+#error "usec_interval_irq_handler should be defined, check device_cfg.h!"
+#endif
+void usec_interval_irq_handler(void)
+{
+    us_ticker_read();
+    us_ticker_irq_handler();
+}
+
+#else // TARGET_MUSCA_A1_NS
+
+void us_ticker_init(void)
```

Comment:
I believe I already asked but again I do not see anything in this file indicating why the implementation is empty. A comment would be worth.

Still, why `TARGET_MUSCA_A1_NS` is providing empty non functional definition?

Reply:
gpio is different, there are comments and implementation empty. Should be the same for serial

Reply:
Empty implementation is here to satisfy linker, but there is an assumption that this code won't be executed. 
Let's make sure this is the case by adding a call to panic/abort and comment.

Reply:
We think that you should not create any empty function.
Code should be with #if DEVICE_USTICKER,
File will not be compiled thx to targets.json setting ?

Reply:
IIRC the are references to this function which are not guarded by any #ifdef guards. @mikisch81, can you double check. 
We agreed we will issue tickets for every such a case. Please make sure we have it covered and add a comment to a ticket to remove this empty implementation

Reply:
OK, so when moving `us_ticker.c` into the NS target I get the following linker errors:
```sh
Error: L6218E: Undefined symbol us_ticker_clear_interrupt (referred from BUILD/ARM_MUSCA_A1_S/ARMC6/mbed-os/hal/mbed_us_ticker_api.o).
Error: L6218E: Undefined symbol us_ticker_disable_interrupt (referred from BUILD/ARM_MUSCA_A1_S/ARMC6/mbed-os/hal/mbed_us_ticker_api.o).
Error: L6218E: Undefined symbol us_ticker_fire_interrupt (referred from BUILD/ARM_MUSCA_A1_S/ARMC6/mbed-os/hal/mbed_us_ticker_api.o).
Error: L6218E: Undefined symbol us_ticker_free (referred from BUILD/ARM_MUSCA_A1_S/ARMC6/mbed-os/hal/mbed_us_ticker_api.o).
Error: L6218E: Undefined symbol us_ticker_init (referred from BUILD/ARM_MUSCA_A1_S/ARMC6/mbed-os/hal/mbed_us_ticker_api.o).
Error: L6218E: Undefined symbol us_ticker_read (referred from BUILD/ARM_MUSCA_A1_S/ARMC6/mbed-os/hal/mbed_us_ticker_api.o).
Error: L6218E: Undefined symbol us_ticker_set_interrupt (referred from BUILD/ARM_MUSCA_A1_S/ARMC6/mbed-os/hal/mbed_us_ticker_api.o).
Finished: 0 information, 2 warning and 7 error messages.
```

When looking into `mbed_us_ticker_api.c` I see that there is a static struct of function pointers - [us_interface](https://github.com/ARMmbed/mbed-os/blob/master/hal/mbed_us_ticker_api.c#L23:L32) which the linker will not throw, therefor even that the secure target does not use us_ticker at all, it needs to resolve those symbols - hence the empty implementation.
I agree this is not the right behaviour, but eventhough the target is using the bare-metal prfofile to build, `hal` folder is still built entirely.
@alzix @donatieng @theotherjimmy @bridadan @jeromecoutant

Reply:
It appears the problem is that [mbed_die()](https://github.com/ARMmbed/mbed-os/blob/master/platform/mbed_board.c#L26-L49) implementation calls [wait_us()](https://github.com/ARMmbed/mbed-os/blob/master/platform/mbed_wait_api_no_rtos.c#L35-L40) which calls `get_us_ticker_data()` which is defined in `mbed_us_ticker_api.c`.
So just guarding `mbed_us_ticker_api.c` with `#if DEVICE_USTICKER` is not enough.

Reply:
Issue #9853 was opened regarding the USTICKER issue.
@ARMmbed/mbed-os-maintainers Can you prioritize this issue, as it is a blocker for porting new secure targets.

Reply:
I am reviewing now. us ticker was always taken as "a must" but anyway there was "USTICKER" in device_has.

cc @kjbracey-arm you might be interested in this issue

Reply:
@0xc0170 `USTICKER` is only defined for `ARM_MUSCA_A1_NS` target, `ARM_MUSCA_A1_S` does not have it in its' `device_has`. The issue is relevant for secure targets which use TF-M and not Mbed-OS.

Reply:
Wasn't this fixed and this empty implementation can be removed?

Reply:
Going to reopen #9853, as I still have dependencies with USTICKER.